src/targets/maven.ts

```diff
@@ -0,0 +1,389 @@
+      };
+    }, {});
+  }
+
+  private getTargetSettings(): Record<TargetSettingType, string> {
+    const settings = targetOptions.map(setting => {
+      return {
+        name: setting,
+        value: this.config[setting],
+      };
+    });
+    return this.reduceConfig(settings);
+  }
+
+  /**
+   * Checks whether the required software to run this target is available
+   * in the system. It assumes the config for this target to be available.
+   * If there's required software missing, raises an error.
+   */
+  private checkRequiredSoftware(): void {
+    this.logger.debug(
+      `Checking if Maven CLI is available on ${this.mavenConfig.mavenCliPath}...`
+    );
+    checkExecutableIsPresent(this.mavenConfig.mavenCliPath);
+    this.logger.debug(
+      `Checking if Gradle CLI is available on ${this.mavenConfig.gradleCliPath}...`
+    );
+    checkExecutableIsPresent(this.mavenConfig.gradleCliPath);
+    this.logger.debug(`Checking if GPG is available in the path...`);
+    checkExecutableIsPresent('gpg');
+  }
+
+  /**
+   * Parses the required target's configuration variables' types from strings to
+   * RegExp.
+   *
+   * To make TS happy, they must be `RegExp` objects. It's not valid to write
+   * `/myRegex/`
+   */
+  private makeRegexpFromConfiguration(): void {
+    this.config.androidDistDirPattern = new RegExp(
+      this.config.androidDistDirPattern
+    );
+    this.config.androidFileSearchPattern = new RegExp(
+      this.config.androidFileSearchPattern
+    );
+  }
+
+  /**
+   * Publishes current Java and Android distributions.
+   * @param version New version to be released.
+   * @param revision Git commit SHA to be published.
+   */
+  public async publish(_version: string, revison: string): Promise<void> {
+    await this.createUserGradlePropsFile();
+    await this.upload(revison);
+    await this.closeAndRelease();
+  }
+
+  /**
+   * Creates the required user's `gradle.properties` file.
+   *
+   * If there's an existing one, it's overwritten.
+   * TODO: control when it's overwritten with an option.
+   */
+  private async createUserGradlePropsFile(): Promise<void> {
+    await fsPromises.writeFile(
+      join(this.getGradleHomeDir(), GRADLE_PROPERTIES_FILENAME),
+      // Using `` instead of string concatenation makes all the lines but the
+      // first one to be indented to the right. To avoid that, these lines
+      // shouldn't have that much space at the beginning, something the linter
+      // doesn't agree with (and the code would be harder to read).
+      `mavenCentralUsername=${this.mavenConfig.MAVEN_CENTRAL_USERNAME}\n` +
+        `mavenCentralPassword=${this.mavenConfig.MAVEN_CENTRAL_PASSWORD}`
+    );
+  }
+
+  /**
+   * Retrieves the Gradle Home path.
+   *
+   * See https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_environment_variables
+   *
+   * @returns the gradle home path.
+   */
+  public getGradleHomeDir(): string {
+    if (process.env.GRADLE_USER_HOME) {
```

Comment:
Why not use `checkEnvForPrerequisite` like above?

Reply:
Because it's not a prerequisite and there aren't any alternative env vars. If it's not defined, Craft should use another path. `checkEnvForPrerequisite` requires the variable to be defined, and raises an error if not.